New ui autumation tests with App Center integration

trwalke opened the pull request:
Increasing UI Automation mobile coverage by adding a silent test along with ADFS V3 and V4 tests.
Enabling app center.

Commit:
Updated ADAL xamarin app.
Added silent UI Automation test

Commit:
Resolved runtime issue with missing assemblies.
More improvements to dev app

Commit:
Merge remote-tracking branch 'origin/dev' into trwalkeNewUIAutumationTests

Commit:
Adding Silent flow test.

Improved ui test infrastructure

Commit:
Updating constants location

Commit:
Adding ADFS v4 and V3 tests

henrik-me commented:
> {

there are lots of duplicate code in this file... pls. refactor #Closed

---
Refers to: adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:11 in 6a33855. [](commit_id = 6a338558b8ff496ff6cca42c35eebbeb0e623b62, deletion_comment = False)

henrik-me commented on adal/automation/WinFormsAutomationApp/WinFormsAutomationApp.csproj:
```diff
@@ -103,8 +103,8 @@
     </PackageReference>
-    <PackageReference Include="Newtonsoft.Json" Version="10.*">
-      <PrivateAssets>All</PrivateAssets>
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2">
```
>11.0.2" [](start = 57, length = 7)

in the two other places you moved to v6. Any reason you are not using the same version everywhere? #Closed

trwalke replied:
The two other places have a hard requirement for v6. this project does not so it can be the latest version #Closed

henrik-me commented:
Great to see the test additions. Are these test cases planned for MSAL as well?  #Closed

henrik-me commented on core/tests/Test.Microsoft.Identity.Core.UIAutomation/Infrastructure/ITestController.cs:
```diff
@@ -46,6 +46,11 @@ public interface ITestController
+        /// <summary>
+        /// Dismisses the keyboard
+        /// </summary>
+        void DismissKeyboard();
```
>void DismissKeyboard(); [](start = 8, length = 23)

I did not see why this is needed in the pr description, but I can see that you have added the usage in several places in the test cases. Why is this needed? #Closed

trwalke replied:
the android keyboard can sometimes hide buttons from the UI automation framework preventing it from clicking on buttons.  #Closed

henrik-me replied:
Kindly ensure this is added to the remarks section of the xml comments

---
In reply to: [216706040](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r216706040) [](ancestors = 216706040)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -31,6 +31,8 @@ public static void AcquireTokenTest(ITestController controller)
             //Clear Cache
             controller.Tap("clearCache");
 
+            controller.Tap("Log");
```
>controller.Tap("Log"); [](start = 12, length = 22)

is this on purpose? reason? #Closed

trwalke replied:
just a test. Removed #Closed

henrik-me replied:
I still see it

---
In reply to: [216706120](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r216706120) [](ancestors = 216706120)

henrik-me replied:
still around afaiks?

---
In reply to: [217874814](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r217874814) [](ancestors = 217874814,216706120)

henrik-me commented on adal/automation/WinFormsAutomationApp/WinFormsAutomationApp.csproj:
```diff
@@ -103,8 +103,8 @@
-    <PackageReference Include="Newtonsoft.Json" Version="10.*">
-      <PrivateAssets>All</PrivateAssets>
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2">
+      <PrivateAssets>all</PrivateAssets>
```
>all [](start = 21, length = 3)

All ? #Closed

trwalke replied:
im not sure, this is added automatically by VS #Closed

bgavrilMS commented on adal/automation/WinFormsAutomationApp/WinFormsAutomationApp.csproj:
```diff
@@ -103,8 +103,8 @@
     </PackageReference>
-    <PackageReference Include="Newtonsoft.Json" Version="10.*">
-      <PrivateAssets>All</PrivateAssets>
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2">
```
I remember I had to move to 10.* because we were getting NuGet warnings. Please make sure we're not getting NuGet warnings :) #Closed

bgavrilMS commented on adal/devApps/XFormsApp/AppConstants.cs:
```diff
@@ -0,0 +1,35 @@
+    public static class AppConstants
+    {
+        //Applications
+        public static readonly string UiAutomationTestClientId = "3c1e0e0d-b742-45ba-a35e-01c664e14b16";
```
Use 'const' instead, it's like 'static readonly' but cannot be initialized from the ctor. #Closed

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -158,15 +212,22 @@ void LogCallback(LogLevel level, string message, bool containsPii)
         }
 
-        private async void AcquireTokenSilentButton_Clicked(object sender, EventArgs e)
+        async void AcquireTokenButton_Clicked(object sender, EventArgs e)
```
1. Is it intentioanal that you made this internal?
2. Consider keeping visibility modifiers explicit because it makes the code easier to read.  #Closed

bgavrilMS commented on adal/devApps/XFormsApp/AppConstants.cs:
```diff
@@ -0,0 +1,35 @@
+        static AppConstants()
+        {
+            Applications = new Dictionary<string, string>();
+            Applications.Add("Ui Test App", UiAutomationTestClientId);
```
Could you add a comment on why we need 2 separate applications ? And perhaps change the string "MSID Lab 4" to somehting more meaningful (.e.g "App Targetting ADFS v2" or whatever Lab4 is doing?) #Closed

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -34,6 +40,235 @@ public static void AcquireTokenTest(ITestController controller)
+            //i0116 = UPN text field on AAD sign in endpoint
+            controller.EnterText("i0116", user.Upn, true);
+            //idSIButton9 = Sign in button
+            controller.Tap("idSIButton9", true);
```
You can always use constants for these strings so that you won't need all these comments.  #Closed

bgavrilMS commented on core/tests/Test.Microsoft.Identity.Core.UIAutomation/Test.Microsoft.Identity.Core.UIAutomation.csproj:
```diff
@@ -32,7 +32,7 @@
     <PackageReference Include="Xamarin.UITest" Version="2.2.4" />
     <PackageReference Include="NUnitTestAdapter" Version="2.1.1" />
-    <PackageReference Include="Newtonsoft.Json" Version="9.*" />
+    <PackageReference Include="Newtonsoft.Json" Version="6.0.8" />
```
Same comment as before - please leave as is unless you have a sepcific reason to use that old version of the lib. Currently we have no build warnings except for the "Cannot build without a MAC" false positive, so we should aim to mantain 0 warnings. #Closed

trwalke replied:
Hard requirement by automation framework to use this version #Closed

bgavrilMS reviewed (commented):
Looks good, some small changes

Commit:
Merge remote-tracking branch 'origin/dev' into trwalkeNewUIAutumationTests

Commit:
Refactored ADAL UIAutomation tests

henrik-me commented:
@trwalke : any updates on this PR? #Closed

trwalke commented:
@henrik-me Yes, however, I am currently blocked on this as the MSAL project is not running the automation test in visual studio. Im trying to figure out what the issue is with help from Bogdan. This issue is reproducable on other machines as well. #Closed

Commit:
Merge remote-tracking branch 'origin/dev' into trwalkeNewUIAutumationTests

Commit:
Adding MSAL equivalent tests.
Refactoring tests to make them more parameterized.

trwalke commented:
@henrik-me This PR includes a workaround to allow the MSAL tests to run. At the moment, the MSAL ui automation project does not allow the tests to run. I have moved them to the ADAL project for the moment until I can figure out a solution to this problem.  #Closed

bgavrilMS commented on adal/devApps/XFormsApp/AppConstants.cs:
```diff
@@ -0,0 +1,38 @@
+            Resources.Add("SharePoint", SharePoint);
+        }
+
+        public static Dictionary<string, string> Applications { get; set; }
```
Consider naming dictionaries in such a way that you mention both key and value, e.g. LabelToResourceUriMap #Closed

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -300,13 +331,46 @@ private async void AcquireTokenSilentWithBroker()
+            switch (Device.RuntimePlatform)
+            {
+                case Device.iOS:
+                    AcquireTokenWithBroker();
```
you can fall through the android case, see https://stackoverflow.com/questions/174155/switch-statement-fallthrough-in-c #Closed

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,152 @@
+        public static void AcquireTokenTest(ITestController controller)
+        {
+            //Get User from Lab
+            var user = controller.GetUser(
```
seems that getting a user and clearing cache are repeated many times, can be consolidated  #Closed

trwalke replied:
#Resolved

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,152 @@
+        public static void AcquireTokenADFSvXInteractiveMSALTest(ITestController controller, FederationProvider federationProvider, bool isFederated)
+        {
+            //Get User from Lab
+            var user = controller.GetUser(
```
Are all these tests new or have just been copied here for mobile test usage? If not new, can the code be reused from desktop tests? #Closed

trwalke replied:
Cant reuse code from desktop tests because the UIs for the desktop apps and mobile apps are all different. #Closed

bgavrilMS reviewed (approved):
LGTM, left some small comments

Commit:
Refactoring to resolve comments

henrik-me commented:
> {

nice!

---
In reply to: [416755769](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#issuecomment-416755769) [](ancestors = 416755769)

---
Refers to: adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:11 in 6a33855. [](commit_id = 6a338558b8ff496ff6cca42c35eebbeb0e623b62, deletion_comment = False)

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,152 @@
+
+            //Clear Cache
+            controller.Tap("Cache");
+            controller.Tap("clearCache");
```
could be consolidated into one and re-user elsewhere #Closed

trwalke replied:
This has been consolidated into prepareForAuthentication() #Closed

henrik-me commented:
great refactoring. still a few comments to address.  #Closed

Commit:
Updating XML comments for DismissKeyboard()

Commit:
Enabling appcenter on the ADAL Xamarin android dev app

Commit:
adding missing test tags.
renaming tests

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs:
```diff
@@ -39,9 +39,54 @@ public void InitializeBeforeTest()
         /// </summary>
         [Test]
-        public void AcquireTokenTest()
+        public void AcquireTokenInteractiveTestADAL()
```
Any reason you are not calling the tests the same across libraries? Wouldn't that make things simpler when moving towards the unified app? #Closed

Commit:
filtering tests for appcenter testing

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            var user = prepareForAuthentication(controller);
+
+            SetInputData(controller, UIAutomationAppV2, "User.Read");
```
>"User.Read" [](start = 56, length = 11)

test scope const? #Pending

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,134 @@
+
+            SetInputData(controller, UIAutomationAppV2, "User.Write");
+
+            controller.Tap("acquireTokenSilent");
```
>"acquireTokenSilent" [](start = 27, length = 20)

const for id's for various controls on the page #Pending

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,134 @@
+        private static IUser prepareForAuthentication(ITestController controller)
+        {
+            //Clear Cache
+            controller.Tap("Cache");
```
>"Cache" [](start = 27, length = 7)

see related comments to const for id's on various controls on the form/page (several similar to this one). I suggest you create a new PR with that or ensure that is part of the unified test app work. #Pending

trwalke commented on adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs:
```diff
@@ -47,7 +47,7 @@ public void AcquireTokenInteractiveTestADAL()
         /// Runs through the standard acquire token silent flow
         /// </summary>
-        [Test]
+        //[Test]
```
please ignore, this is just for app center testing #Closed

henrik-me replied:
? ok... 

---
In reply to: [219612532](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r219612532) [](ancestors = 219612532)

trwalke commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -28,6 +28,20 @@ public static void AcquireTokenInteractiveTest(ITestController controller)
 		{
             var user = prepareForAuthentication(controller);
 
+            // Create a string array with the lines of text
```
please ignore, this is just for app center testing #Closed

henrik-me commented:
still tiny stuff to be done

---
In reply to: [421528501](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#issuecomment-421528501) [](ancestors = 421528501)

Commit:
Refactoring to reduce copied code.
Adding const strings for IDs.

trwalke commented:
@henrik-me removed appcenter testing code and refactored #Closed

trwalke committed:
Update CoreMobileMSALTests.cs

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -14,38 +14,132 @@ namespace Test.ADAL.NET.UIAutomation
+        private const string WebPasswordID = "i0118";
+        private const string WebSubmitID = "idSIButton9";
+        private const string TestResultID = "testResult";
+        private const string TestResultSuccsesfulMessage = "Result: Success";
```
put all of these in a helper class which both the ADAL and the MSAL test can reference #Closed

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -0,0 +1,145 @@
+        private static void VerifyResult(ITestController controller)
+        {
+            //Test results are put into a label that is checked for messages
+            Assert.IsTrue(controller.GetText(TestResultID).Contains(TestResultSuccsesfulMessage));
```
>Assert.IsTrue(controller.GetText(TestResultID).Contains(TestResultSuccsesfulMessage)); [](start = 12, length = 86)

remember to clear the text field before running tests, otherwise one successful test will result in all tests being successful.  #Closed

trwalke replied:
This is cleared before every authentication call by the test app itself #Closed

henrik-me commented:
Looks great. a few new minor comments.

---
In reply to: [423655746](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#issuecomment-423655746) [](ancestors = 423655746,421528501)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/CoreMobileMSALTests.cs:
```diff
@@ -31,6 +31,8 @@ public static void AcquireTokenTest(ITestController controller)
             //Clear Cache
             controller.Tap("clearCache");
 
+            controller.Tap("Log");
```
>controller.Tap("Log"); [](start = 12, length = 22)

Is this needed?

henrik-me reviewed (changes requested):
:clock1:

Commit:
Enabling App Center

Commit:
Revert "Enabling App Center"

This reverts commit 3f314393b0ec682194da6ab65a24d11d7a175fe1.

henrik-me commented on msal/tests/dev apps/XForms/XForms/AcquirePage.xaml.cs:
```diff
@@ -172,6 +178,7 @@ private async Task OnAcquireClickedAsync(object sender, EventArgs e)
         {
             try
             {
+                acquireResponseTitleLabel.Text = "Result:";
```
>"Result:" [](start = 49, length = 9)

const (used several places) #Pending

henrik-me commented on msal/tests/dev apps/XForms/XForms/AcquirePage.xaml.cs:
```diff
@@ -160,7 +161,12 @@ private async Task OnAcquireSilentlyClickedAsync(object sender, EventArgs e)
+                var resText = ToString(res);
+
+                if (resText.Contains("AccessToken"))
+                    acquireResponseTitleLabel.Text = "Result: Success";
```
>"Result: Success" [](start = 53, length = 17)

const. used several places #Pending

henrik-me commented:
>             //App.UIParent = new UIParent(this, true);

uncomment here, an comment above, instead of replacing? Assuming the UI tests currently only works with embedded web view? what is the plan for supporting the system browser again? #Closed

---
Refers to: msal/tests/dev apps/XForms/XForms.Android/MainActivity.cs:56 in 6bba5f6. [](commit_id = 6bba5f680b8b36ba468dea9447f9e74a61709c26, deletion_comment = False)

henrik-me commented:
merge conflicts #Closed

Commit:
Implementing client secret for keyvault Auth

Commit:
Updating tests to nunit 3

Commit:
Moving MSAL ui tests back to the msal project.
refactoring mobile automation tests.
Adding shared constants file for ADAL and MSAL ui tests.

Commit:
Merge remote-tracking branch 'origin/dev' into trwalkeNewUIAutumationTests

# Conflicts:
#	adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs
#	core/tests/Test.Microsoft.Identity.Core.UIAutomation/Infrastructure/FederationProvider.cs
#	msal/tests/Test.MSAL.NET.UIAutomation/CoreMobileMSALTests.cs

Commit:
Merge remote-tracking branch 'origin/trwalke/AppCenter' into trwalkeNewUIAutumationTests

# Conflicts:
#	adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs
#	adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidMSALTests.cs
#	core/tests/Test.Microsoft.Identity.Core.UIAutomation/Infrastructure/KeyVaultConfiguration.cs

Commit:
Cleaning up merge errors

Commit:
Moving datafile to proper location

trwalke commented:
The data.txt file is a placeholder for app center. Data will be written to it during build time on VSTS #Pending

MarkZuber commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -110,8 +113,13 @@ private async Task<string> AuthenticationCallbackAsync(string authority, string
+                    break;
                 case KeyVaultAuthenticationType.UserCredential:
-                    authResult = await authContext.AcquireTokenAsync(resource, _config.ClientId, new UserCredential());
+                    authResult = await authContext.AcquireTokenAsync(resource, _config.ClientId, new UserCredential(""));
```
string.Empty instead of "" #Closed

MarkZuber commented on msal/tests/dev apps/XForms/XForms.Android/MainActivity.cs:
```diff
@@ -47,13 +47,13 @@ protected override void OnCreate(Bundle bundle)
             LoadApplication(new App());
 
-            App.UIParent = new UIParent(this);
+            //App.UIParent = new UIParent(this);
```
is this dead code?  if so, remove. #Closed

trwalke replied:
it is not, we uncomment it when we want to enable the android app to use system webview #Closed

MarkZuber replied:
Could you please add a comment above this to specify the intention then? #Closed

MarkZuber commented on msal/tests/dev apps/XForms/XForms/AcquirePage.xaml.cs:
```diff
@@ -160,7 +163,12 @@ private async Task OnAcquireSilentlyClickedAsync(object sender, EventArgs e)
-                acquireResponseLabel.Text = ToString(res);
+                var resText = ToString(res);
+
+                if (resText.Contains("AccessToken"))
```
please wrap all conditional expressions (even one liners) with brackets { }

so:

if (resText.Contains("AccessToken"))
{
    acquireResponseTitleLabel.Text = SuccessfulResult;
} #Closed

MarkZuber reviewed (changes requested):
...

Commit:
implementing app center workaround for vsts

Commit:
addressing PR comments

MarkZuber reviewed (approved):
Please add the comment mentioned in my notes and otherwise looks great.  thanks!

Commit:
Updating coments for system web view in MSAL

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
 
-            //Verify result. Test results are put into a label
-            Assert.IsTrue(controller.GetText("testResult") == "Success: True");
+				Thread.Sleep(1000);
```
>Thread.Sleep(1000); [](start = 4, length = 19)

argh... I get really scared about Thread.Sleeps in the code.... pls. ensure you have a work item tracking getting rid of this.

bgavrilMS replied:
Actually it seems this is just a polling mechanism so a form of waiting is fine. Would be better if the app would use async / await and then you could wait without blocking.

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs:
```diff
@@ -20,6 +20,19 @@ class XamarinDroidADALTests
         Platform platform;
         ITestController xamarinController;
 
+        UserQueryParameters DefaultUser
```
nit: can be moved to a testuser helper? #Closed

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -84,7 +87,19 @@ public class KeyVaultSecretsProvider
             _config = new KeyVaultConfiguration();
-            _config.AuthType = KeyVaultAuthenticationType.ClientCertificate;
+
+            //The data.txt is a place holder for the keyvault secret. it will only be written to during build time when testing appcenter
```
>//The data.txt is a place holder for the ke [](start = 12, length = 43)

pls. ensure the file is properly cleaned up after the test run #Pending

bgavrilMS replied:
This is an "unmanaged" resource, so please implement the IDisposable pattern here.

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/LabServiceApi.cs:
```diff
@@ -68,6 +68,11 @@ private IEnumerable<IUser> GetUsersFromLab(UserQueryParameters query)
 
+            if (String.IsNullOrWhiteSpace(result))
+            {
+                throw new Exception("No lab user with specified parameters exists");
```
>throw new Exception [](start = 16, length = 19)

you should create a specific exception for this. generally you should never use the system.exception #Closed

bgavrilMS replied:
InvalidOperationException is the "common" exception type when you don't want to have a new exception (.e.g. only throw it once). Probably worth having a specific one for these tests. #Closed

henrik-me commented on msal/tests/dev apps/XForms/XForms.Android/MainActivity.cs:
```diff
@@ -47,13 +47,14 @@ protected override void OnCreate(Bundle bundle)
-            // To activate embedded webview, remove '//' from line 56 below, 
-            // and comment out line 50 above -> App.UIParent = new UIParent(Xamarin.Forms.Forms.Context as Activity);
+            // To activate embedded webview, remove '//' from line 57 below, 
+            // and comment out line 51 above -> App.UIParent = new UIParent(Xamarin.Forms.Forms.Context as Activity);
```
keeping these numbers in check... is difficult. code should be refactored to make this easier to understand and setup.... ideally this should be controlled by a setting in the test app. #Pending

trwalke replied:
I'll resolve in a follow up PR #Pending

henrik-me replied:
ok.

---
In reply to: [224643193](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r224643193) [](ancestors = 224643193)

henrik-me commented on msal/tests/dev apps/XForms/XForms.Android/MainActivity.cs:
```diff
@@ -47,13 +47,14 @@ protected override void OnCreate(Bundle bundle)
+            // and comment out line 51 above -> App.UIParent = new UIParent(Xamarin.Forms.Forms.Context as Activity);
 
-            //App.UIParent = new UIParent(this, true);
+            App.UIParent = new UIParent(this, true);
```
>App.UIParent = new UIParent(this, true); [](start = 12, length = 40)

you now have both UIParent setters enabled.... really should rather be a meaningful if statement or as mentioned... a setting in the UI.

henrik-me replied:
this one is still set 2 times.... please see above

---
In reply to: [223935008](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r223935008) [](ancestors = 223935008)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/XamarinDroidMSALTests.cs:
```diff
@@ -42,6 +42,19 @@ class XamarinMSALDroidTests
         Platform platform;
         ITestController xamarinController;
 
+        UserQueryParameters DefaultUser
```
commented elsewhere. this thing can be in a helper .... and potentially the other user types can as well... #Closed

henrik-me commented:
added a few new comments. but really this looks great, super improvements! You can chose to create follow up tasks to each of these items and let me know if that is your approach or you can address asks... pls. stop by for a chat

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -300,13 +331,44 @@ private async void AcquireTokenSilentWithBroker()
+                    AcquireTokenWithBroker();
+                    break;
+                default:
+                    this.result.Text = "UWP does not support broker. Use iOS or Android.";
```
1. It is safer to treat all cases of the enum and throw in the default. This is because in the future we may add to the enum and the behaviour of this logic may not be the one expected. 
 #Closed

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -300,13 +331,44 @@ private async void AcquireTokenSilentWithBroker()
+                    AcquireTokenSilentWithBroker();
+                    break;
+                default:
+                    this.result.Text = "UWP does not support broker. Use iOS or Android.";
```
As above, better throw.

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+        /// Runs through the standard acquire token silent flow
+        /// </summary>
+        /// <param name="controller">The test framework that will execute the test interaction</param>
+        public static void AcquireTokenSilentTest(ITestController controller, UserQueryParameters userParams)
```
All the methods in this class are static and all take in a test controller and user params. This is a code smell that in fact this class should not be static and ITestController is a member of the class. Probably UserQueryParameters as well.

trwalke replied:
What would be the benefit of making the class concrete? Also, if I make ITestController a member of the class, each test that uses this class would need to first set the ITestController before the test code is executed anyway. Thus, I thought it was better to simply pass the ITestController directly into the function that contains the test code, saving a few lines of code overall.

trwalke replied:
This is meant to be more of a helper class for the automation tests. I have renamed it so that it makes more sense. Please see my latest changes. Perhaps it will change your opinion?

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+        /// Runs through the standard acquire token interactive flow
+        /// </summary>
+        /// <param name="controller">The test framework that will execute the test interaction</param>
+        public static void AcquireTokenADFSvXInteractiveTest(ITestController controller, bool isFederated, UserQueryParameters userParams)
```
isFederated is not used

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+        private static IUser prepareForAuthentication(ITestController controller, UserQueryParameters userParams)
+        {
+            //Navigate to second page
+            controller.Tap(UiTestConstants.secondPageID);
```
style - SecondPageId (uppercase not camel case)

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+            return controller.GetUser(userParams);
+        }
+
+        private static void SetInputData(ITestController controller, string ClientID, string Resource)
```
params are camelCase

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+            switch (user.FederationProvider)
+            {
+                case FederationProvider.AdfsV4:
+                    passwordInputID = UiTestConstants.AdfsV4WebPasswordID;
```
I am a bit confused here - why do we need passwords ? Is this for the U/P flow? The class and the methods do not make this clear.

trwalke replied:
This is meant to be more of a helper class for the automation tests. I have renamed it so that it makes more sense. This function is a generic sign in pattern for interactive flows on mobile platforms. Thus, it is not specific to any test.

bgavrilMS commented on adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs:
```diff
@@ -37,37 +40,131 @@ namespace Test.ADAL.NET.UIAutomation
+            //Thus this method will check the result once a second for 10 seconds.
+            bool done = false;
+            int attempts = 0;
+            int maximumAttempts = 20;
```
See if you can encapsulate a retry strategy into some retry helper and pull those constants somewhere. For inspiration https://alastaircrabtree.com/implementing-a-simple-retry-pattern-in-c/

Commit:
Adding specific exceptions for test result failuses and lab user failures.
Implementing dispose on keyvaultSecretsProvider.
Implementing retry pattern for test result verification.
Refactoring.

henrik-me commented on adal/tests/Test.ADAL.NET.UIAutomation/ADALMobileTestHelper.cs:
```diff
@@ -0,0 +1,185 @@
+                catch (ResultVerificationFailureException ex)
+                {
+                    if (attempts == UiTestConstants.maximumResultCheckRetryAttempts)
+                        throw new Exception("Could not Verify test result", ex);
```
>throw new Exception [](start = 24, length = 19)

why not Assert.Fail ? #Closed

trwalke replied:
after a second look, there is no reason not to. I changed it #Closed

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:
```diff
@@ -0,0 +1,182 @@
+    /// <summary>
+    /// Contains the core test functionality that will be used by Android and iOS tests
+    /// </summary>
+    public static class MSALMobileTestHelper
```
>MSALMobileTestHelper [](start = 24, length = 20)

what is the reason the test helpers can't be in common? #Closed

trwalke replied:
no, I was move some the functionality to core. #Closed

henrik-me commented:
one last comment imo, which needs to be adressed, imo

---
In reply to: [428436335](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#issuecomment-428436335) [](ancestors = 428436335)

bgavrilMS commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -119,5 +147,27 @@ private async Task<string> AuthenticationCallbackAsync(string authority, string
+            }
+
+            // Disposing unmanaged resources.
+            _keyVaultClient.Dispose();
```
>Dispose [](start = 28, length = 7)

This is good, if a class has a member that implements IDisposable, it should implement IDisposable as well and call Dispose on that member. However, you should also delete that data.txt file in your Dispose - this is the "unmanaged" resource this class is holding. #Pending

trwalke replied:
According to the documentation for File.ReadAllText()
"The file handle is guaranteed to be closed by this method, even if exceptions are raised."
With that, I believe we can safely assume that the resource is disposed properly.

https://docs.microsoft.com/en-us/dotnet/api/system.io.file.readalltext?redirectedfrom=MSDN&view=netframework-4.7.2#System_IO_File_ReadAllText_System_String_ #Pending

trwalke replied:
Also, the data within the file will be cleared via VSTS task #Pending

henrik-me replied:
This ask was about ensuring the file is deleted from the device it's created on and where it may be sent to. Is it deleted both from the build and from the device which runs the tests?

henrik-me replied:
as said this is not about disposing the returned object from reading the string. but the text file itself (imo)

---
In reply to: [224925931](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#discussion_r224925931) [](ancestors = 224925931)

trwalke replied:
Updated comments to make the disposal of this file more clear.

Commit:
Refactoring test helper methods into a core test helper.
Fixed issue where the keyvaultSecretsProvider tried to authenticate multiple times.

trwalke commented:
@henrik-me PR is updated. I also found a bug with the keyvault secrets provider and resolved it.

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -54,6 +58,12 @@ public class KeyVaultSecretsProvider
 
         private readonly string keyVaultThumbPrint = "440A5BE6C4BE2FF02A0ADBED1AAA43D6CF12E269";
 
+        bool disposed = false;
```
>disposed [](start = 13, length = 8)

updated style _disposed

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -54,6 +58,12 @@ public class KeyVaultSecretsProvider
 
+        bool disposed = false;
+
+        private readonly string dataFile = "data.txt";
```
>dataFile [](start = 32, length = 8)

updated style: _dataFile ... should probably be _dataFileName

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/KeyVaultSecretsProvider.cs:
```diff
@@ -110,14 +150,41 @@ private async Task<string> AuthenticationCallbackAsync(string authority, string
+            }
+
+            // Disposing unmanaged resources.
+            _keyVaultClient.Dispose();
```
>_keyVaultClient.Dispose(); [](start = 12, length = 26)

recommended pattern is to check if _keyVaultClient is null, if not call dispose and set to null... that ensures dispose can be called multiple times

henrik-me commented on core/tests/Test.Microsoft.Identity.Core.UIAutomation/CoreMobileTestHelper.cs:
```diff
@@ -0,0 +1,104 @@
+                catch (ResultVerificationFailureException ex)
+                {
+                    if (attempts == UiTestConstants.maximumResultCheckRetryAttempts)
+                        throw new Exception("Could not Verify test result", ex);
```
>throw new Exception("Could not Verify test result", ex); [](start = 24, length = 56)

before you moved it you mentioned this could be an assert instead... did you change your mind?

trwalke replied:
Yes, I realized that there wouldn't be much value in using an exception rather than Assert.Fail().

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/ILabService.cs:
```diff
@@ -31,6 +31,6 @@ namespace Test.Microsoft.Identity.LabInfrastructure
     public interface ILabService
     {
-        IEnumerable<IUser> GetUsers(UserQueryParameters query);
+        IUser GetUsers(UserQueryParameters query);
```
If you go by returning just one, you should also change the method name to be GetUser

henrik-me reviewed (approved):
:shipit:

henrik-me commented:
discussed a few things in person. a few more changes are coming and then this will be good to go.

---
In reply to: [429476021](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1235#issuecomment-429476021) [](ancestors = 429476021)

Commit:
Refactoring member names.
Downgrading nunit back to 2.6.4 for appcenter compatability.
adding relevant comments.

Commit:
Refactoring/Renaming files

Commit:
Renaming projects to avoid test issues in VSTS

Commit:
Merge remote-tracking branch 'origin/dev' into trwalkeNewUIAutumationTests

# Conflicts:
#	adal/devApps/AdalDesktopTestApp/Program.cs
#	adal/devApps/XFormsApp.Droid/XFormsApp.Droid.csproj
#	adal/devApps/XFormsApp/XFormsApp.csproj
#	adal/tests/Test.ADAL.NET.UIAutomation/CoreMobileADALTests.cs
#	core/tests/Test.Microsoft.Identity.LabInfrastructure/Test.Microsoft.Identity.LabInfrastructure.csproj
#	msal/tests/dev apps/XForms/XForms/AcquirePage.xaml.cs

Commit:
resolving build error
